File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        long long int n=height.size(),lmax=height[0],rmax=height[n-1],lpoi=1,rpoi=n-2,sol=0;
        while(lpoi<=rpoi){
            if(height[lpoi]>=lmax){
                lmax=height[lpoi++];
            }
            else if(height[rpoi]>=rmax){
                rmax=height[rpoi--];
            }
            else if(height[lpoi]<=lmax && lmax<rmax){
                sol+=lmax-height[lpoi++];
            }
            else{
                sol+=rmax-height[rpoi--];
            }   
        }
        return sol;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        sol=0
        n=len(height)
        lmax=[0]*len(height)
        rmax=[0]*len(height)
        lmax[0]=height[0]
        rmax[len(height)-1]=height[n-1]
        for i in range(1,len(height)):
            lmax[i]=max(lmax[i-1],height[i])
        for i in range (n-2,0,-1):
            rmax[i]=max(rmax[i+1],height[i])
        for i in range(1,len(height)-1):
            sol+=max(0,min(lmax[i],rmax[i])-height[i])
        return sol

        